rfcs/0001-restructure.md

```diff
@@ -0,0 +1,271 @@
+# Restructuring PHP Buildpacks
+
+## Proposal
+
+The RFC proposes a new set of smaller buildpacks that, by themselves provides
+more limited functionality, but in composition provides a powerful way to build
+various kinds of images related to PHP.
+
+## Motivation
+
+1. Reducing the responsibility of each buildpack will help to make them easier
+   to understand and maintain. We will end up with more buildpacks, each with
+   simpler implementations.
+
+1. Align buildplan philosophy with our other buildpacks. We have moved away
+   from the buildpack unilaterally marking its layers as `build` or `launch`.
+   Buildpacks should interact with each other via meaningful `build` and
+   `launch` requirements. This builds a clean separation between the concerns
+   of building and running an application image.
+
+In the current state of the world, we have:
+
+- a [php-dist](github.com/paketo-buildpacks/php-dist) buildpack that provides
+  php.
+- a [php-composer](github.com/paketo-buildpacks/php-composer) that provides
+  composer and installs dependencies.
+- a [php-web](github.com/paketo-buildpacks/php-web) that takes care of
+  everything related to serving php on a web server.
+
+## Buildpacks
+
+The PHP family of buildpacks aims to serve the following functions:
+* Build an image to run HTTPD web server with php
+* Build an image to run Nginx web server with php
+* Build an image to run Built-in php web server
+* Build an image to run FPM[<sup>1</sup>](#note-1) process manager
+* Optionally use composer as application level package manager
+* Optionally use memcached/redis as session handler[<sup>3</sup>](#note-3)
+
+The new structure would include the following buildpacks in addition to the existing
+Apache HTTPD Server and Nginx Server buildpacks[<sup>2</sup>](#note-2):
+
+* **php-dist**:
+  Installs the [`php`](https://www.php.net) distribution, making it available on the `$PATH`
+  * provides: `php`
+  * requires: none
+
+  This distribution must include popular modules like memcached, mongodb,
+  mysqli, openssl, phpiredis, zlib etc; and a suitable `php.ini`. Downstream
+  buildpacks are expected to obtain info (like extension dir) through the
+  [`php-config`](https://www.php.net/manual/en/install.pecl.php-config.php)
+  executable whenever possible.
+
+* **composer**:
+  Installs [`composer`](https://getcomposer.org), a dependency manager for PHP
+  and makes it available on the `$PATH`
+  * provides: `composer`
+  * requires: none
+
+* **composer-install**:
```

Comment:
Are the names on `composer` and `composer-install` clear? On my first read-through I interpreted them to have the opposite meaning. In that `composer-install` would install composer and `composer` would run composer to do things.

Thinking about it a bit more I get why the current set exists. Composer will install composer itself, and `composer-install` runs the `composer install` command, but that seems to limit things. We can/will want to run other `composer` commands.

My vote would be to flip them around and have `composer-install` handle installing composer, and have `composer` handle running composer.

Reply:
The names were picked in keeping with the convention widely used in paketo buildpacks. [node](https://github.com/paketo-buildpacks/yarn) buildpack provides `yarn` to the image where as [yarn-install](https://github.com/paketo-buildpacks/yarn-install) and [npm-install](https://github.com/paketo-buildpacks/npm-install) buildpacks have `yarn install` and `npm install` as their principal build commands. These buildpacks also run other commands (like `npm ci`). We have the same pattern in {[bundler](https://github.com/paketo-buildpacks/bundler), [bundle-install](https://github.com/paketo-buildpacks/bundle-install)}, {[pip, pip-install](https://github.com/paketo-community/python/blob/6cf9d677e20286db982a2d38d0a356f361bf6397/rfcs/0001-restructure.md)} etc. Do the names make sense in the context of this naming convention?

Reply:
It still sounds backward IMHO, but if that's how the others are structured 👍. Consistency is important as well.